File: src/main/java/io/transwarp/logparser/util/FileParser.java
package io.transwarp.logparser.util;

import io.transwarp.logparser.filter.*;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

/**
 * Author: stk
 * Date: 2018/4/2
 */
public class FileParser {
    private static final List<String> patterns = Loader.loadDatePattern();
    private List<Filter> filters;

    public List<Record> parseFile(Path path) {
        initFilters();
        System.out.println("Parsing file: " + path);
        List<Record> records = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(path.toString()))) {
            String line;
            while ((line = reader.readLine()) != null) if (isBegin(line)) break; // Begin from first record.
            List<String> lines = new ArrayList<>();
            lines.add(line);

            while ((line = reader.readLine()) != null) {
                if (isBegin(line)) {
                    Record record = new Record(lines);
                    if (filter(record)) records.add(record);
                    lines.clear();
                }
                lines.add(line);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return records;
    }

    private void initFilters() {
        filters = new ArrayList<>();
        filters.add(new TimeFilter("20000101 00:00:00,000", "20500101 00:00:00,000"));
        filters.add(new LevelFilter());
//        filters.add(new ExceptionFilter());
//        filters.add(new DuplicationFilter());
    }

    private boolean isBegin(String line) {
        return patterns.parallelStream().anyMatch(s -> Pattern.compile(s).matcher(line).find());
    }

    private boolean filter(Record record) {
        return filters.stream().dropWhile(filter -> filter.filter(record)).count() == 0;
    }
}

File: src/main/scala/io/transwarp/logparser/util/FileParser.scala
package io.transwarp.logparser.util

import java.io.File

import io.transwarp.logparser.filter._

import scala.io.Source

/**
  * Author: stk
  * Date: 2018/4/4
  */
object FileParser {
  private val pattern: List[String] = Source.fromResource("pattern").getLines.map(Converter.convertDateToRegex).toList

  def parseFile(file: File): List[Record] = {
    val filters = List(new TimeFilter("20000101 00:00:00,000", "20500101 00:00:00,000"),
      new LevelFilter)

    println("Parsing file: " + file)
    var records: List[Record] = List()
    var fileLines = Source.fromFile(file).getLines.toList
    fileLines = fileLines.dropWhile(!isBeginLine(_))
    var lines: List[String] = List()

    val filter = (lines: List[String]) => if (lines.nonEmpty) {
      val record = new Record(lines)
      if (filters.dropWhile(_.filter(record)).isEmpty) records = records :+ record
    }

    for (line <- fileLines) {
      if (isBeginLine(line)) {
        filter(lines)
        lines = List()
      }
      lines = lines :+ line
    }
    filter(lines)
    records
  }

  private def isBeginLine(line: String): Boolean = pattern.map(_.r.findFirstIn(line)).exists(_.isDefined)
}
